### PR TITLE
rgw: clean up ordered list

### PR DESCRIPTION
**Pull request by ivancich**
Ordered bucket listing uses parallel vectors to manage moving through the listings from individual shards to produce a single ordered list. Parallel vectors are error prone. This replaces them with a vector of structs to coherently manage the state of where we are within each shard's listing.

Fixes: https://tracker.ceph.com/issues/44160

### CODE REVIEW

**Comment by ivancich** [comment]
please unstale

File: src/rgw/rgw_rados.cc [hunk 1]
```diff
@@ -8316,9 +8360,9 @@ int RGWRados::cls_bucket_list_ordered(RGWBucketInfo& bucket_info,
 
     // refresh the candidates map
     candidates.erase(candidates.begin());
-    if (++vcurrents[pos] != vends[pos]) { // note: pre-increment
-      candidates[vcurrents[pos]->first] = pos;
-    } else if (list_results[pos].is_truncated) {
+    if (! tracker.advance().at_end()) { // note: pre-increment
```
Comment (cbodley) [inline src/rgw/rgw_rados.cc]: the comment `// note: pre-increment` no longer applies here

Comment (cbodley) [inline src/rgw/rgw_rados.cc]: the comment `// note: pre-increment` no longer applies here

Comment (ivancich) [reply src/rgw/rgw_rados.cc]: Fixed. Thanks!

**Comment by ivancich** [comment]
teuthology QA results:

rgw/verify:
http://pulpito.ceph.com/ivancich-2020-02-18_18:29:22-rgw-wip-clean-up-ordered-list-distro-basic-smithi/
re-run of failure: http://pulpito.ceph.com/ivancich-2020-02-18_21:40:46-rgw-wip-clean-up-ordered-list-distro-basic-smithi/

rgw/(non-)verify:
http://pulpito.ceph.com/ivancich-2020-02-19_18:28:40-rgw-wip-clean-up-ordered-list-distro-basic-smithi/
re-run of failures w/o specifying centos: http://pulpito.ceph.com/ivancich-2020-02-20_16:21:34-rgw-wip-clean-up-ordered-list-distro-basic-smithi/

**Pull request merged**